firmware/library/L1_Drivers/onboardled.hpp

```diff
@@ -7,14 +7,99 @@
 #pragma once
 
 #include <cstdint>
+#include "L1_Drivers/gpio.hpp"
+#include "L2_Utilities/macros.hpp"
 
 class OnBoardLedInterface {
  public:
-           virtual void Initialize()                                 = 0;
-           virtual void LedOn(uint8_t led_number)                    = 0;
-           virtual void LedOff(uint8_t led_number)                   = 0;
-           virtual void LedSet(uint8_t led_number, bool on)          = 0;
-           virtual void LedToggle(uint8_t led_number)                = 0;
-           virtual void LedSetAll(uint8_t value)                     = 0;
-           virtual uint8_t LedGetValues(void)                        = 0;
+  enum class LightState : bool
+  {
+      kOn = true,
+      kOff = false
+  };
+           virtual void Initialize()                                      = 0;
+           virtual void On(uint8_t led_number)                            = 0;
+           virtual void Off(uint8_t led_number)                           = 0;
+           virtual void Set(uint8_t led_number, LightState state)         = 0;
+           virtual void Toggle(uint8_t led_number)                        = 0;
+           virtual void SetAll(uint8_t value)                             = 0;
+           virtual uint8_t GetStates(void)                                = 0;
+};
+
+class OnBoardLed: public OnBoardLedInterface
+{
+ public:
+  // Initialize takes the array of Gpios, sets each one to an output, and
+  // then turns all of the leds by setting the output high.
+  void Initialize(void) override
+    {
+      for (uint8_t i = 0; i < 4; i++)
+      {
+        led[i].SetAsOutput();
+        led[i].SetHigh();
+      }
+    }
+
+  void On(uint8_t led_number) override
+  {
+    SJ2_ASSERT_WARNING(led_number > 3,
```

Comment:
Sorry for the back and forth, but looking at this after a nights sleep, I realize that this will cause a hard-fault or out of array access if you merely `ASSERT_WARNING`. Assert warning allows the code to continue be emits a warning message. `SJ2_ASSERT_FATAL` would be better since it will halt the program at that point.